solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -246,6 +245,138 @@ public boolean canBeRemoved() {
+     */
+    private long activeCollectionCount() {
+      return statefulWatchesByCollectionName.entrySet().stream()
+          .filter(
+              (Entry<String, StatefulCollectionWatch> entry) ->
+                  entry.getValue().currentState != null)
+          .count();
+    }
+
+    private Set<String> watchedCollections() {
+      return Collections.unmodifiableSet(statefulWatchesByCollectionName.keySet());
+    }
+
+    private Set<Entry<String, StatefulCollectionWatch>>
+        watchedCollectionEntries() {
+      return Collections.unmodifiableSet(statefulWatchesByCollectionName.entrySet());
+    }
+
+    /**
+     * Updates the latest observed DocCollection (state) of the {@link StatefulCollectionWatch} if
+     * the collection is being watched
+     *
+     * @param collection the collection name
+     * @param newState the new DocCollection (state) observed
+     * @return whether an active watch exists for such collection
+     */
+    private boolean updateDocCollection(String collection, DocCollection newState) {
+      StatefulCollectionWatch finalWatch =
+          statefulWatchesByCollectionName.computeIfPresent(
+              collection,
+              (col, watch) -> {
+                DocCollection oldState = watch.currentState;
+                if (oldState == null && newState == null) {
+                  // OK, the collection not yet exist in ZK or already deleted
+                } else if (oldState == null) {
+                  if (log.isDebugEnabled()) {
+                    log.debug("Add data for [{}] ver [{}]", collection, newState.getZNodeVersion());
+                  }
+                  watch.currentState = newState;
+                } else if (newState == null) {
+                  log.debug("Removing cached collection state for [{}]", collection);
+                  watch.currentState = null;
+                } else { // both new and old states are non-null
+                  int oldCVersion =
+                      oldState.getPerReplicaStates() == null
+                          ? -1
+                          : oldState.getPerReplicaStates().cversion;
+                  int newCVersion =
+                      newState.getPerReplicaStates() == null
+                          ? -1
+                          : newState.getPerReplicaStates().cversion;
+                  if (oldState.getZNodeVersion() < newState.getZNodeVersion()
+                      || oldCVersion < newCVersion) {
+                    watch.currentState = newState;
+                    if (log.isDebugEnabled()) {
+                      log.debug(
+                          "Updating data for [{}] from [{}] to [{}]",
+                          collection,
+                          oldState.getZNodeVersion(),
+                          newState.getZNodeVersion());
+                    }
+                  }
+                }
+                return watch;
+              });
+      return finalWatch != null;
+    }
```

Comment:
```suggestion
     * @return whether the state has changed for the watched collection
     */
    private boolean updateDocCollection(String collection, DocCollection newState) {
      AtomicBoolean stateHasChanged = new AtomicBoolean(false);
      statefulWatchesByCollectionName.computeIfPresent(
          collection,
          (col, watch) -> {
            DocCollection oldState = watch.currentState;
            if (oldState == null && newState == null) {
              // OK, the collection not yet exist in ZK or already deleted
            } else if (oldState == null) {
              if (log.isDebugEnabled()) {
                log.debug("Add data for [{}] ver [{}]", collection, newState.getZNodeVersion());
              }
              watch.currentState = newState;
            } else if (newState == null) {
              log.debug("Removing cached collection state for [{}]", collection);
              watch.currentState = null;
            } else { // both new and old states are non-null
              int oldCVersion =
                  oldState.getPerReplicaStates() == null
                      ? -1
                      : oldState.getPerReplicaStates().cversion;
              int newCVersion =
                  newState.getPerReplicaStates() == null
                      ? -1
                      : newState.getPerReplicaStates().cversion;
              if (oldState.getZNodeVersion() < newState.getZNodeVersion()
                  || oldCVersion < newCVersion) {
                watch.currentState = newState;
                if (log.isDebugEnabled()) {
                  log.debug(
                      "Updating data for [{}] from [{}] to [{}]",
                      collection,
                      oldState.getZNodeVersion(),
                      newState.getZNodeVersion());
                }
              }
            }
            stateHasChanged.set(oldState != watch.currentState);
            return watch;
          });

      return stateHasChanged.get();
    }
```

Reply:
@HoustonPutman Your implementation makes perfect sense as it does reflect whether the state has been updated, which if the existing [`updateWatchedCollection` before our change](https://github.com/apache/solr/blob/7339c37ad5893ad003e3d73fa72007e1ddd60747/solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java#L1951)  does behave as described in method comment, then it will be all good!

Unfortunately, the original `updateWatchedCollection` basically has 2 cases which violate what it describes as `returns true if the state has changed`:
1. If both new and old states are null (which could happen if either the collection is not yet created or is removed), it would still return `true`
2. If the old state has version number >= the new state, even though there's no update, it would still return `true`

The 2nd case in particular could be problematic, as described in the comment:
```
// no change to state, but we might have been triggered by the addition of a
          // state watcher, so run notifications
```
Basically the method is trying to behave differently because it has knowledge of certain behavior requirement of the caller, such reverse dependency is usually not desirable.

And my original attempt was to remove such reverse dependency by simply observing what does the return value actually mean after I simplified the logic to mimic the exact return value of the original `updateWatchedCollection`. My proposed change turns out to be equally confusing 😓